pkg/machine/connection/connection.go

```diff
@@ -85,37 +83,51 @@ func UpdateConnectionIfDefault(rootful bool, name, rootfulName string) error {
 	})
 }
 
-func RemoveConnections(names ...string) error {
-	return config.EditConnectionConfig(func(cfg *config.ConnectionsFile) error {
-		for _, name := range names {
-			if _, ok := cfg.Connection.Connections[name]; ok {
-				delete(cfg.Connection.Connections, name)
-			} else {
-				return fmt.Errorf("unable to find connection named %q", name)
-			}
+func RemoveConnections(machines map[string]bool, names ...string) error {
+	var dest config.Destination
+	var service string
 
-			if cfg.Connection.Default == name {
-				cfg.Connection.Default = ""
-			}
-		}
-		for service := range cfg.Connection.Connections {
-			cfg.Connection.Default = service
-			break
-		}
-		return nil
-	})
+	if err := config.EditConnectionConfig(func(cfg *config.ConnectionsFile) error {
+		return setNewDefaultConnection(cfg, &dest, &service, names...)
+	}); err != nil {
+		return err
+	}
+
+	rootful, ok := machines[service]
+	if dest.IsMachine && ok {
+		return UpdateConnectionIfDefault(rootful, service, service+"-root")
+	}
```

Comment:
This is racy, you are opening the connections file twice (and thus lock/unlock twice)
You must keep the entire logic in one `config.EditConnectionConfig()` call otherwise on concurrent removal you might end up adding a connection back after it was removed between both calls here

Reply:
Thanks for pointing that out, should be fixed now!